helper/consts/plugin_types.go

```diff
@@ -0,0 +1,40 @@
+package consts
+
+import "fmt"
+
+var PluginTypes = []PluginType{PluginTypeCredential, PluginTypeSecrets, PluginTypeDatabase}
+
+type PluginType int
+
+const (
+	PluginTypeUnknown PluginType = iota
+	PluginTypeCredential
+	PluginTypeSecrets
+	PluginTypeDatabase
+)
+
+func (p PluginType) String() string {
+	switch p {
+	case PluginTypeSecrets:
+		return "secret"
+	case PluginTypeCredential:
+		return "auth"
+	case PluginTypeDatabase:
+		return "database"
+	default:
+		return "unknown"
+	}
+}
+
+func ParsePluginType(pluginType string) (PluginType, error) {
+	switch pluginType {
+	case "secret":
+		return PluginTypeSecrets, nil
+	case "auth":
+		return PluginTypeCredential, nil
+	case "database":
+		return PluginTypeDatabase, nil
+	default:
+		return PluginTypeUnknown, fmt.Errorf("%s is not a supported plugin type", pluginType)
```

Comment:
Should this function return an error at all, I wonder? Returning a value _and_ and error is atypical. It also seems like since Unknown is a valid defined type, that it isn't really an error condition at all. There isn't any more information being returned to the caller in this error response, and their logic can check for `PluginTypeUnknown` as necessary.

Reply:
What would you think about stripping `PluginTypeUnknown` and just returning -1 there? It would look like this: https://play.golang.org/p/H1kmvX6Kkaq

I had only added unknown because I thought I couldn't return -1. I actually really like returning an error because it is definitely an erroneous condition if the plugin type can't be parsed. If that function returned only `PluginTypeUnknown`, then every calling function in my code would need to check for unknown and then build its own error, essentially moving a very repetitive piece _out_ of the function. Also, IMHO it's more idiomatic to check for an error than for a particular const that may come back and it seems a little like a gotcha.

Reply:
I take that back! As I'm working on some of the other code changes requested here, I'm finding that the unknown type is actually useful. I still do think that it should be an error if you can't parse something. If you feel strongly, I can update it though.

Reply:
I think returning `PluginTypeUnknown` is fine in here (since we can't really return a nil value) as long as we consider `PluginTypeUnknown` a non-valid case and take the correct action (e.g. error out) appropriately if we see that. Since we should always be error-checking first, this _should_ be a non-issue.

Reply:
It think the main issue is that the strong coupling between the value and error might cause confusion. Such as:

```
pluginType, err :=  ParsePluginType(pluginTypeStr)
if err != nil {
	return err
}
if pluginType == PluginTypeUnknown {
	// do special thing with "unknown" plugin
}
```
This looks reasonable, but the latter block will never be hit. If type Unknown has a non-error interpretation, then it shouldn't be returned with an error. OTOH if Unknown always means error, then we're returning two error values for one error state, effectively.

Reply:
Also, parsing time works exactly like this: https://golang.org/pkg/time/#Parse. If there's an error state, it returns an error and the default time, expecting the caller not to use it.

Reply:
I think I was conflating the usage pattern of this with that of BackendType, which does have more widespread use of Unknown in a non-error sense. I've looked a bit more at PluginTypeUnknown use and now agree that it's OK as is. 👍

Reply:
Hm, that's a good point, I will go through the code and think about that more. Also, FWIW, that last comment of mine was posted before GH revealed your comment, I think they're having issues today. So it makes it appear that my comment was in response to yours, but it was actually in response to the earlier comments. :-)

Reply:
I went through the code and saw that now that `PluginTypeUnknown` is a valid case in some sections, you're right, it would be wise to add it. Here's the commit: https://github.com/hashicorp/vault/pull/5536/commits/f8c71dd999eeec1940616473c9448854a1f595a6. Also, I added a verbose note for maintenance unrelatedly.

Reply:
LGTM!